File: engine/src/elements/Background.ts
import Element from '../Element'
import Size from '../Size'
import BoxConstraints from '../BoxConstraints'

interface BackgroundProps {
	color: string
}

class Background extends Element<BackgroundProps> {
	layout(constraints: BoxConstraints) {
		return new Size({
			width: constraints.maxWidth,
			height: constraints.maxHeight
		})
	}

	paint() {
		const layer = this.document.getLayer()
		const ctx = layer.canvas.getContext('2d')
		ctx.save()
		ctx.translate(this.absPosition.left, this.absPosition.top)
		ctx.fillStyle = this.props.color
		ctx.fillRect(0, 0, this.size.width, this.size.height)
		ctx.restore()
	}
}

export default Background

File: core/src/elements/background.cpp
#include "background.hpp"
#include <iostream>

namespace aardvark::elements {

Background::Background(SkColor color, bool is_repaint_boundary)
    : Element(is_repaint_boundary, /* size_depends_on_parent */ true),
      color(color){};

void Background::set_props(SkColor color) {
  this->color = color;
  document->change_element(this);
};

Size Background::layout(BoxConstraints constraints) {
  return Size{
      constraints.max_width /* width */, constraints.max_height /* height */
  };
};

void Background::paint(bool is_changed) {
  SkPaint paint;
  paint.setColor(color);
  auto layer = document->get_layer();
  document->setup_layer(layer, this);
  SkIRect rect{0, 0, size.width, size.height};
  layer->canvas->drawIRect(rect, paint);
};

}  // namespace aardvark::elements
